docs/proposals/20210301-enhancement_of_YurtHub_caching_ability.md

```diff
@@ -0,0 +1,375 @@
+   // map[string]interface{}
+   // children.
+   Object map[string]interface{}
+}
+```
+
+#### 2.1 Design detail
+
+##### Add unstructuredNegotiatedSerializer
+
+On the basis of the existing NegotiatedSerializer increase a new unstructuredNegotiatedSerializer, used to decode the structure of the unregistered Custom Resources.
+
+```go
+type unstructuredNegotiatedSerializer struct {
+	scheme  *runtime.Scheme
+	typer   runtime.ObjectTyper
+	creator runtime.ObjectCreater
+}
+
+// SerializerManager is responsible for managing *rest.Serializers
+type SerializerManager struct {
+	// NegotiatedSerializer is used for obtaining encoders and decoders for multiple
+	// supported media types.
+	NegotiatedSerializer runtime.NegotiatedSerializer
+    //TO-ADD: use to decode the CR
+    unstructuredNegotiatedSerializer runtime.NegotiatedSerializer
+}
+```
+
+##### Select the appropriate serializer based on GVK information
+
+When cache resources, get the GVK information of resource based on the  GVR in `request.URL` and `RESTMapper` (which is used to check whether the GVK is in the scheme according to the GVR information). An implementation of `RESTMapper` is as follows:
+
+```go
+type DefaultRESTMapper struct {
+	defaultGroupVersions []schema.GroupVersion
+
+	resourceToKind       map[schema.GroupVersionResource]schema.GroupVersionKind
+	kindToPluralResource map[schema.GroupVersionKind]schema.GroupVersionResource
+	kindToScope          map[schema.GroupVersionKind]RESTScope
+	singularToPlural     map[schema.GroupVersionResource]schema.GroupVersionResource
+	pluralToSingular     map[schema.GroupVersionResource]schema.GroupVersionResource
+}
+```
+
+If the resource's GVK has been registered, use the original serializer, otherwise use the serializer returned by `unstructuredNegotiatedSerializer` .
+
+![unstructure](../img/img-20210301001.png)
+
+### 3. Abstract the logic of processing objects into a separate processing layer
+
+The object processing logic is divided into two layers: filtering layer and processing layer. For each GVR, the corresponding filtering, adding, updating and deleting logic is set and managed.
+
+#### 3.1 Design detail
+
+The following is a flowchart for decoding, processing, and saving runtime.Object locally:
+
+![new flow](../img/img-20210323001.png)
+
+In the cacheManager, add `handlerLayer` to process object:
+
+```go
+type cacheManager struct {
+	sync.RWMutex
+	storage           StorageWrapper
+	serializerManager *serializer.SerializerManager
+    //add handlerLayer to process object
+	handlerLayer 	  HandlerLayer
+	cacheAgents       map[string]bool
+}
+```
+
+Followings are definitions of `HandlerLayer` interface and its related  structs. It can use the `AddHandler(...)` functions to add the handlers for different GVR.
+
+```go
+type HandlerLayer interface {
+	AddHandler(gvr schema.GroupVersionResource, handler ResourceEventHandler) error
+	GetHandler(gvr schema.GroupVersionResource) ResourceEventHandler
```

Comment:
How about add more than one ResourceEventHandler for one gvr, so GetHandler will return a slice of ResourceEventHandler.

Reply:
Thanks for advice. In consideration of function call order, one handler chain will be stored for each GVR.